File: internal/src/main/java/dev/mvmo/sicklang/internal/object/hash/HashObject.java
package dev.mvmo.sicklang.internal.object.hash;

import dev.mvmo.sicklang.internal.object.ObjectType;
import dev.mvmo.sicklang.internal.object.SickObject;
import dev.mvmo.sicklang.internal.object.hashkey.HashKey;

import java.util.Map;
import java.util.stream.Collectors;

public record HashObject(Map<HashKey, Entry> pairs) implements SickObject {

    @Override
    public String inspect() {
        return String.format("{%s}", pairs.values().stream()
                .map(entry -> String.format("%s: %s", entry.key().inspect(), entry.value().inspect()))
                .collect(Collectors.joining(", ")));
    }

    @Override
    public ObjectType objectType() {
        return ObjectType.HASH;
    }

    public static record Entry(SickObject key, SickObject value) {
    }

}

File: internal/src/main/java/dev/mvmo/sicklang/internal/object/hash/HashObject.kt
package dev.mvmo.sicklang.internal.`object`.hash

import dev.mvmo.sicklang.internal.`object`.ObjectType
import dev.mvmo.sicklang.internal.`object`.SickObject
import dev.mvmo.sicklang.internal.`object`.hashkey.HashKey

class HashObject(val pairs: Map<HashKey, Entry>) : SickObject {

    override fun inspect() =
        "{${pairs.map { it.value }.joinToString { "${it.key}: ${it.value}" }}}"

    override fun objectType() =
        ObjectType.HASH

    data class Entry(val key: SickObject, val value: SickObject)

}